src/Workspaces/Core/Portable/FindSymbols/FindReferences/MetadataUnifyingEquivalenceComparer.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal sealed class MetadataUnifyingEquivalenceComparer : IEqualityComparer<ISymbol>
+    {
+        public static readonly IEqualityComparer<ISymbol> Instance = new MetadataUnifyingEquivalenceComparer();
+
+        private MetadataUnifyingEquivalenceComparer()
+        {
+        }
+
+        public bool Equals(ISymbol x, ISymbol y)
+        {
+            // If either symbol is from source, then we must do stricter equality. Consider this:
+            //
+            //     S1 <-> M <-> S2     (where S# = source symbol, M = some metadata symbol)
+            //
+            // In this case, imagine that both the comparisons denoted by <-> were done with the
+            // SymbolEquivalenceComparer, and returned true. If S1 and S2 were from different projects,
+            // they would compare false but transitivity would say they must be true. Another way to think
+            // of this is any use of a source symbol "poisons" the comparison and requires it to be stricter.
+            if (x == null || y == null || IsInSource(x) || IsInSource(y))
```

Comment:
I'm somewhat terrified of this.  Does this mean that in *normal* i.e. non-xproj cases, we might run into a case where, due to retargetting, we see two source symbols that now are considered different.  i.e.:

P1 <-- P2  (p2 has a reference to p1, but retargets p1 because of hte metadata references).

Now, might we hit the same symbol from P1 twice?  once from P1 itself as we're walking/cascading, and once when we see it wrapped as a retargetting symbol in P2?  Are we going to then potentially get lots of results that look identical to the user?